nova/scheduler/filters/shard_filter.py

```diff
@@ -148,18 +215,58 @@ def host_passes(self, host_state, spec_obj):
         if self._ALL_SHARDS in shards:
             LOG.debug('project enabled for all shards %(project_shards)s.',
                       {'project_shards': shards})
-            return True
         elif host_shard_names & set(shards):
             LOG.debug('%(host_state)s shard %(host_shard)s found in project '
                       'shards %(project_shards)s.',
                       {'host_state': host_state,
                        'host_shard': host_shard_names,
                        'project_shards': shards})
-            return True
         else:
             LOG.debug('%(host_state)s shard %(host_shard)s not found in '
                       'project shards %(project_shards)s.',
                       {'host_state': host_state,
                        'host_shard': host_shard_names,
                        'project_shards': shards})
             return False
+
+        if not utils.request_is_resize(spec_obj):
+            # K8S orchestrators are only creating or deleting nodes,
+            # therefore we shouldn't infer with resize/migrate requests.
+            return self._host_passes_k8s(host_state, host_shard_names,
+                                         spec_obj, k8s_instances)
+
+        return True
+
+    def _host_passes_k8s(self, host_state, host_shard_names, spec_obj,
+                         k8s_instances):
+        """Instances of a K8S cluster must end up on the same shard.
+        The K8S cluster is identified by the metadata or tags set
+        by the orchestrator (Gardener or Kubernikus).
+        """
+        if not k8s_instances:
+            # There are no instances in the cluster on this availability
+            # zone, yet. We allow any shard in this AZ for the first instance.
+            return True
+
+        def _is_hana_flavor(flavor):
+            return flavor.name.startswith(HANA_PREFIX)
+
+        def _is_same_category(instance, flavor):
+            """Check whether instance is from the flavor's family."""
+            if _is_hana_flavor(flavor):
+                return _is_hana_flavor(instance.flavor)
+            return True
+
+        siblings = [i for i in k8s_instances
+                    if _is_same_category(i, spec_obj.flavor)]
+
+        if not siblings:
+            # This is the first instance of this particular type (HANA).
+            # We allow any shard in this AZ for the first instance.
+            return True
+
+        k8s_hosts = set([i.host for i in siblings])
```

Comment:
We compute this for every host even though it's not host-dependent. Could we pass `k8s_hosts` instead of `k8s_instances` into the method and compute `k8s_hosts` in `filter_all()`?